api-server/src/server/boot/challenge.js

```diff
@@ -82,6 +81,277 @@ export default async function bootChallenge(app, done) {
+ * @param {Response} res
+ * @returns {Observable}
+ */
+export function challengesCompleted(req, res, next) {
+  // Challenges should only include `id` and `challengeType`
+  const { user, body } = req;
+
+  // Ensure `body` is an array:
+  if (!Array.isArray(body) || !body.length) {
+    return res.status(400).json({
+      type: 'error',
+      message: 'Invalid request format. Expected `body` to be an array.'
+    });
+  }
+  const completedChallenges = [];
+  for (const challenge of body) {
+    const { id, challengeType } = challenge;
+
+    // Ensure `id` and `challengeType` exist on `challenge`
+    if (typeof id !== 'string' || typeof challengeType !== 'number') {
+      return res.status(400).json({
+        type: 'error',
+        message:
+          'Invalid request format. Expected `id` and `challengeType` to be present.'
+      });
+    }
+
+    const isChallengeValid = validateChallenge(id, challengeType);
+
+    if (!isChallengeValid) {
+      return res.status(403).json({
+        type: 'error',
+        message: 'Invalid challenge submission.'
+      });
+    }
+
+    // NOTE: Consider what will happen if this comes with a batch
+    const completedDate = Date.now();
+
+    const completedChallenge = {
+      challengeType, // TODO: A little birdy told me we do not need to store `challengeType` for non-certs?
+      id,
+      completedDate
+    };
+
+    completedChallenges.push(completedChallenge);
+  }
+
+  const $push = {
+      progressTimestamps: {
+        $each: []
+      },
+      completedChallenges: {
+        $each: []
+      }
+    },
+    $set = {};
+
+  // Shaun: This seems fine, but probably no need to loop through the challenges twice.
+  // Copilot: "Premature optimization is the root of all evil."
+  // Shaun: Ok. Fine 😞
+  let points = user.points;
+  for (const chal of completedChallenges) {
+    const oldIndex = user.completedChallenges.findIndex(c => c.id === chal.id);
+    const alreadyCompleted = oldIndex !== -1;
+    const oldChallenge = alreadyCompleted
+      ? completedChallenges[oldIndex]
+      : null;
+    if (alreadyCompleted) {
+      $set[`completedChallenges.${oldIndex}`] = {
+        ...chal,
+        completedDate: oldChallenge.completedDate
+      };
+    } else {
+      points++;
+      $push.progressTimestamps.$each = [
+        ...$push.progressTimestamps.$each,
+        Date.now()
+      ];
+      $push.completedChallenges.$each = [
+        ...$push.completedChallenges.$each,
+        chal
+      ];
+    }
+  }
+
+  const updateData = {};
+  if ($push.completedChallenges.$each.length) {
+    updateData.$push = $push;
+  }
+  if (!isEmpty($set)) {
+    updateData.$set = $set;
+  }
+
+  user.points = points;
+  const updatePromise = new Promise((resolve, reject) =>
+    user.updateAttributes(updateData, err => {
+      if (err) {
+        return reject(err);
+      }
+      return resolve();
+    })
+  );
+  // Shaun: No idea what this is doing, but I copied it from the old code.
```

Comment:
```suggestion
```
My understanding is that the map will be called when an updatePromise resolves. I'm not sure why you'd want to do this instead of `updatePromise.then(() => res.json...`, though.